Compiler warning when enabling userspace, sockets and speed optimization

**Describe the bug**
When migrating my application from Zephyr v3.0.0 to v3.1.0 I got a compile warning.
I was able to reproduce it on a Zephyr sample application, see below.

**To Reproduce**
Steps to reproduce the behavior:
1. Add the following lines to `zephyr/samples/userspace/prod_consumer/prj.conf`:
```
CONFIG_SPEED_OPTIMIZATIONS=y
CONFIG_ENTROPY_STM32_RNG=y
CONFIG_NETWORKING=y
CONFIG_NET_SOCKETS=y
```
2. `west build -d build zephyr/samples/userspace/prod_consumer -b stm32f769i_disco`
3. See error

**Expected behavior**
I expect no compile warnings

**Impact**
Annoyance

**Logs and console output**
```
helge@x1:~/tmp/nucleus$ west build -d build zephyr/samples/userspace/prod_consumer -b stm32f769i_disco
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /home/helge/tmp/nucleus/zephyr/samples/userspace/prod_consumer
-- Found Python3: /usr/bin/python3.10 (found suitable exact version "3.10.4") found components: Interpreter 
-- Cache files will be written to: /home/helge/.cache/zephyr
-- Zephyr version: 3.1.99 (/home/helge/tmp/nucleus/zephyr)
-- Found west (found suitable version "0.13.1", minimum required is "0.7.1")
-- Board: stm32f769i_disco
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.13.1 (/home/helge/zephyr-sdk-0.13.1)
-- Found dtc: /home/helge/zephyr-sdk-0.13.1/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6")
-- Found toolchain: zephyr 0.13.1 (/home/helge/zephyr-sdk-0.13.1)
-- Found BOARD.dts: /home/helge/tmp/nucleus/zephyr/boards/arm/stm32f769i_disco/stm32f769i_disco.dts
-- Generated zephyr.dts: /home/helge/tmp/nucleus/build/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /home/helge/tmp/nucleus/build/zephyr/include/generated/devicetree_unfixed.h
-- Generated device_extern.h: /home/helge/tmp/nucleus/build/zephyr/include/generated/device_extern.h
-- Including generated dts.cmake file: /home/helge/tmp/nucleus/build/zephyr/dts.cmake
Parsing /home/helge/tmp/nucleus/zephyr/Kconfig
Loaded configuration '/home/helge/tmp/nucleus/zephyr/boards/arm/stm32f769i_disco/stm32f769i_disco_defconfig'
Merged configuration '/home/helge/tmp/nucleus/zephyr/samples/userspace/prod_consumer/prj.conf'
Configuration saved to '/home/helge/tmp/nucleus/build/zephyr/.config'
Kconfig header saved to '/home/helge/tmp/nucleus/build/zephyr/include/generated/autoconf.h'
-- The C compiler identification is GNU 10.3.0
-- The CXX compiler identification is GNU 10.3.0
-- The ASM compiler identification is GNU
-- Found assembler: /home/helge/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
CMake Warning at /home/helge/tmp/nucleus/zephyr/CMakeLists.txt:1809 (message):
  __ASSERT() statements are globally ENABLED


-- Configuring done
-- Generating done
-- Build files have been written to: /home/helge/tmp/nucleus/build
-- west build: building application
[1/225] Preparing syscall dependency handling

[2/225] Generating include/generated/version.h
-- Zephyr version: 3.1.99 (/home/helge/tmp/nucleus/zephyr), build: zephyr-v3.1.0-442-gfd622650c3d0
[93/225] Building C object zephyr/CMakeFiles/zephyr.dir/subsys/net/lib/sockets/sockets.c.obj
In file included from /home/helge/tmp/nucleus/zephyr/subsys/net/lib/sockets/sockets.c:20:
/home/helge/tmp/nucleus/build/zephyr/include/generated/syscalls/zsock_accept_mrsh.c: In function 'z_mrsh_zsock_accept':
/home/helge/tmp/nucleus/zephyr/include/zephyr/syscall_handler.h:349:23: warning: 'addrlen_copy' may be used uninitialized in this function [-Wmaybe-uninitialized]
  349 |  Z_SYSCALL_VERIFY_MSG(arch_buffer_validate((void *)ptr, size, write) \
      |                       ^~~~~~~~~~~~~~~~~~~~
[203/225] Linking C executable zephyr/zephyr_pre0.elf

[214/225] Linking C executable zephyr/zephyr_pre1.elf

[225/225] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      158872 B         2 MB      7.58%
            SRAM:       52008 B       384 KB     13.23%
            DTCM:       12448 B       128 KB      9.50%
          SDRAM1:          0 GB        16 MB      0.00%
        IDT_LIST:          0 GB         2 KB      0.00%

```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: zephyr-sdk-0.13.1
 - Zephyr v3.1.0  (also fails on latest main branch)

--- a/subsys/net/lib/sockets/sockets.c
+++ b/subsys/net/lib/sockets/sockets.c
@@ -639,9 +639,10 @@ static inline int z_vrfy_zsock_accept(int sock, struct sockaddr *addr,
 
 	Z_OOPS(addrlen && z_user_from_copy(&addrlen_copy, addrlen,
 					   sizeof(socklen_t)));
-	Z_OOPS(addr && Z_SYSCALL_MEMORY_WRITE(addr, addrlen_copy));
+	Z_OOPS(addr && Z_SYSCALL_MEMORY_WRITE(addr, addrlen ? addrlen_copy : 0));
 
-	ret = z_impl_zsock_accept(sock, (struct sockaddr *)addr, &addrlen_copy);
+	ret = z_impl_zsock_accept(sock, (struct sockaddr *)addr,
+				  addrlen ? &addrlen_copy : NULL);
 
 	Z_OOPS(ret >= 0 && addrlen && z_user_to_copy(addrlen, &addrlen_copy,
 						     sizeof(socklen_t)));

